iam_groups_authn/sync.py

```diff
@@ -71,26 +86,24 @@ async def groups_sync(iam_groups, sql_instances, credentials, private_ip=False):
 
     for instance in sql_instances:
         instance_task = asyncio.create_task(get_instance_users(user_service, instance))
-        database_version_task = asyncio.create_task(
-            user_service.get_database_version(
-                InstanceConnectionName(*instance.split(":"))
-            )
+        database_version = await user_service.get_database_version(
```

Comment:
awaiting this here means we are losing some parallelization - the database versions are no longer being gathered as the same point. Consider just moving this check below (either into it's own for loop, or the group iteration)

Reply:
Probably we want the failures to be as contained a possible, but we still want these calls to happen in parallel. So we should schedule the calls first, then await/validate them in a separate loop. Although since we want to push it down to the lowest level, it'll probably look like this:

1. Schedule `get_instance_users` call
2. Schedule `get_database_version` call
3. Schedule `sync_group`, and pass it 1 & 2
3a. each sync_group should probably log if it passes/fails
4. Wait until all of 3 is complete before returning.

Reply:
Added a `sync_group` function and now I gather all the tasks for instance-to-group pairings at once so that they are as contained as possible. (i.e. mapping two groups and two instances, 3 of the pairings can pass successfully and sync while 1 can fail)